### ISSUE TITLE
keras2onnx2ncnn时，flatten层无法支持

### ISSUE THREAD

**Issue by scutzzj** [Tencent/ncnn#1136]
您好，最近在做模型转换的工作，采用的思路是keras2onnx2ncnn，用LeNet做实验。
以下是keras中的模型：
model = Sequential() 
model.add(Conv2D(20, (5, 5), padding="same", activation="relu", input_shape=X_train.shape[1:], name="conv1")) 
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name="pool1"))
model.add(Conv2D(50, (5, 5), padding="same", activation="relu", name="conv2"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name="pool2"))
model.add(Flatten(name="flatten"))
model.add(Dense(500, activation="relu", name="fc1"))
model.add(Dense(num_category, activation="sigmoid", name="fc2"))

从keras能转换到onnx，我查看了onnx的模型，发现特别是flatten层被拆分成多层，如下
  adjusted_input = Transpose[perm = [0, 3, 1, 2]](%conv1_input_01)
  convolution_output = Conv[auto_pad = 'NOTSET', dilations = [1, 1], group = 1, kernel_shape = [5, 5], pads = [2, 2, 2, 2], strides = [1, 1]](%adjusted_input, %W, %B)
  conv1_Relu_0 = Relu(%convolution_output)
  pooling_output = MaxPool[auto_pad = 'VALID', ceil_mode = 0, kernel_shape = [2, 2], strides = [2, 2]](%conv1_Relu_0)
  convolution_output1 = Conv[auto_pad = 'NOTSET', dilations = [1, 1], group = 1, kernel_shape = [5, 5], pads = [2, 2, 2, 2], strides = [1, 1]](%pooling_output, %W1, %B1)
  conv2_Relu_0 = Relu(%convolution_output1)
  pooling_output1 = MaxPool[auto_pad = 'VALID', ceil_mode = 0, kernel_shape = [2, 2], strides = [2, 2]](%conv2_Relu_0)
  pool2_MaxPool_0 = Transpose[perm = [0, 2, 3, 1]](%pooling_output1)
  TFNodes_pool2_MaxPool_0 = Identity(%pool2_MaxPool_0)
  TFNodes_flatten_stack_Unsqueeze__12_0 = Unsqueeze[axes = [0]](%TFNodes_flatten_stack_0_0)
  TFNodes_flatten_Shape_0 = Shape(%TFNodes_pool2_MaxPool_0)
  TFNodes_flatten_Shape__7_0 = Cast[to = 1](%TFNodes_flatten_Shape_0)
  TFNodes_flatten_strided_slice_0 = Slice(%TFNodes_flatten_Shape__7_0, %const_slice__8, %const_slice__9, %const_slice__10)
  TFNodes_flatten_strided_slice__11_0 = Cast[to = 6](%TFNodes_flatten_strided_slice_0)
  TFNodes_flatten_Prod_0 = ReduceProd[axes = [0], keepdims = 0 (%TFNodes_flatten_strided_slice__11_0)
  TFNodes_flatten_stack_Unsqueeze__13_0 = Unsqueeze[axes = [0]](%TFNodes_flatten_Prod_0)
  TFNodes_flatten_stack_Concat__14_0 = Concat[axis = 0](%TFNodes_flatten_stack_Unsqueeze__12_0, %TFNodes_flatten_stack_Unsqueeze__13_0)
  TFNodes_flatten_Reshape__15_0 = Cast[to = 7](%TFNodes_flatten_stack_Concat__14_0)
  TFNodes_flatten_Reshape_0 = Reshape(%TFNodes_pool2_MaxPool_0, TFNodes_flatten_Reshape__15_0)
  transformed_tensor = MatMul(%TFNodes_flatten_Reshape_0, %W2)
  biased_tensor_name = Add(%transformed_tensor, %B2)
  fc1_Relu_0 = Relu(%biased_tensor_name)
  transformed_tensor1 = MatMul(%fc1_Relu_0, %W3)
  biased_tensor_name1 = Add(%transformed_tensor1, %B3)
  fc2_Sigmoid_01 = Sigmoid(%biased_tensor_name1)
  fc2_Sigmoid_0 = Identity(%fc2_Sigmoid_01)

上述模型中TFNodes开头的基本都是flatten层，将该模型用onnx2ncnn.exe转换后，发现无法生成正确的ncnn模型，报错如下：
Unsqueeze not supported yet!
  axes 7
Shape not supported yet!
Cast not supported yet!
  to=1
Cast not supported yet!
  to=6
ReduceProd not supported yet!
  axes 7
  keepdims=0
Unsqueeze not supported yet!
  axes 7
Cast not supported yet!
  to=7
Identity not supported yet!

请问这一问题要如何解决？我的keras模型应该没问题，可以转换成tensorflow的pb模型并正确预测样本。之前试过降低keras版本，发现flatten层依然不支持。
能否在karas中找到替换flatten层的方法？
或者让ncnn能支持上述层？
或者能否提供其他keras2ncnn或者tf2ncnn的路径？
十分感谢！

**Comment by nihui** [comment]
https://github.com/Tencent/ncnn/wiki/practical-pytorch-to-onnx-to-ncnn#simplify-onnx-model

**Comment by nihui** [comment]
ReduceProd added in https://github.com/Tencent/ncnn/commit/04b4b023249865ae01befe4170a873d5f2ec33f8

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx

**Comment by Zejun-Yang** [comment]
> 针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn
> 
> ```shell
> pip install pnnx
> pnnx model.onnx inputshape=[1,3,224,224]
> ```
> 
> 详细参考文档 Detailed reference documentation https://github.com/pnnx/pnnx https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx

泪目
